test/unit/common/msRestRequestTest.js

```diff
@@ -0,0 +1,102 @@
+var should = require('should');
+var sinon = require('sinon');
+
+var msRestRequest;
+var Token = require('../../../lib/common/token');
+
+describe('msRestRequest', function() {
+  before(function(){
+    sinon.stub(Token, 'getToken').yields(null, 'fake-token');
+  });
+  
+  after(function(){
+    Token.getToken.restore();
+  });
+  
+  describe('when the token is not expired and statusCode is not in AZURE_RETRY_ERROR_CODES', function() {
+    
+    before(function() {
+      require('request');
+      require.cache[require.resolve('request')].exports = function(_, callback) {
+        callback(null, {statusCode: 200}, {});
+      };
+      delete require.cache[require.resolve('../../../lib/common/msRestRequest')];
+      msRestRequest = require('../../../lib/common/msRestRequest');
+    });
+    
+    after(function(){
+      delete require.cache[require.resolve('request')];
+    });
+    
+    it('should not exist error', function(done) {
+      msRestRequest.GET('', {}, '', function(err, res, body) {
+        should.not.exist(err);
+        done();
+      });
+    });
+  });
+
+  describe('when the token is not expired, statusCode is in AZURE_RETRY_ERROR_CODES, and retry successfully', function() {
+    
+    var callTimes = 0;
+    
+    before(function() {
+      require('request');
+      require.cache[require.resolve('request')].exports = function(_, callback) {
+        ++callTimes;
+        if (callTimes == 1) {
+          callback(null, {statusCode: 502}, {});
```

Comment:
There are methods in sinon test framework to achieve this nicely `.onFirstCall()` http://sinonjs.org/releases/v1.17.7/stubs/